### PR TITLE
feat: add sorceror + text input

### LINKED ISSUE #157: bug: using meeting.InputType = "text" crashes front end
image

### PR DESCRIPTION
**Pull request by pikulet**
- added sorcerer (v ** d ** lady renamed)
- added text input on the front end (previously it was a copy paste of another input type)
- the text input has a box to type and a button to submit (prevents autosubmitting when player is typing halfway)
- the text input is configurable using "textOptions", e.g. min max length, remove symbols or whitespaces 

**able to die in the middle of the day and curse through filters**
image

**text input with min-max length and option to lowercase or alpha only**
image
the text at the bottom is what has been submitted. the text is the box is lost when you refresh and not submit.

**curses through whitespaces and caps**
image

**effect only lasts one day**

Fixes #157

### CODE REVIEW

File: Games/core/Meeting.js [hunk 1]
```diff
@@ -86,6 +86,7 @@ module.exports = class Meeting {
 
         if (options.inputType)
             this.inputType = options.inputType;
+            this.textOptions = options.textOptions;
```
Comment (pikulet) [inline Games/core/Meeting.js:89]: for passing textoptions to the frontend

File: Games/types/Mafia/roles/cards/CurseWithWord.js [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+
+    constructor(role) {
+        super(role);
+
+        this.meetings = {
+            "Write on Talisman": {
+                states: ["Night"],
+                flags: ["voting"],
+                inputType: "text",
+                textOptions: {
```
Comment (pikulet) [inline Games/types/Mafia/roles/cards/CurseWithWord.js:14]: configurable options for how text input looks like in frontend

File: react_main/src/Constants.jsx [hunk 3]
```diff
@@ -45,6 +45,7 @@ export const RatingThresholds = {
 export const RequiredTotalForStats = 1;
 
 export const MaxGameMessageLength = 240;
+export const MaxTextInputLength = 100;
```
Comment (pikulet) [inline react_main/src/Constants.jsx:48]: value is consistent with data/constants `maxGameTextInputLength`

File: react_main/src/pages/Game/Game.jsx [hunk 4]
```diff
@@ -1699,36 +1699,55 @@ function ActionButton(props) {
...
-            <div
-                className={`btn btn-theme ${votes[props.self] == targetDisplay ? "sel" : ""}`}
-                key={target}
-                onClick={() => onVote(target)}>
-                {targetDisplay}
-            </div>
-        );
-    });
+    function handleOnChange(e) {
+        var textInput = e.target.value;
```
Comment (pikulet) [inline react_main/src/pages/Game/Game.jsx:1713]: use textoptions to add regex filters

File: react_main/src/pages/Game/Game.jsx [hunk 5]
```diff
@@ -1699,36 +1699,55 @@ function ActionButton(props) {
...
+            </div>
+            <textarea
+                value={textData}
+                onChange={handleOnChange} />
+            <div
+                className="btn btn-theme"
+                onClick={handleOnSubmit}>
+                {textOptions.submit || "Submit"}
+            </div>
+            {meeting.votes[self]}
```
Comment (pikulet) [inline react_main/src/pages/Game/Game.jsx:1750]: displays submitted value for user

**Comment by DrSharky** [comment]
Regarding sorceror, if you curse someone with armor (Armored modifier or Knight), it doesn't kill, but it also doesn't break the armor. It actually silences the message and seemingly does nothing else.
Also, an explosive player killed by the curse explodes the sorceror. Not sure if that's intended, but figured I'd mention.

**Comment by pikulet** [comment]
Thanks! So the armour silence is because "this.dominates()" has a nil dereference error when emitting the immune event. Have fixed it by adding "game": this.actor.game as an action property.

**Review by ChaoticNeutralGnome — DISMISSED** [review]
I still think voodoo should have power 2 to go through armor, but it looks good

**Review by starryburst — DISMISSED** [review]

**Review by ChaoticNeutralGnome — APPROVED** [review]

**Review by ChaoticNeutralGnome — APPROVED** [review]

**Pull request merged**